File: Data Structures & Algorithms/is-anagram/submission-3.java
class Solution {
    public boolean isAnagram(String s, String t) {
        //given 2 strings s and t, return true if the strings are naagrams

        //to be an anagram:
            //stirng contains same NUMBER of characters
            //String contains same characters

        //first check if the strings are the same length
        //first condition easily checked
        if(s.length() != t.length()){
            return false;
        }
        //we can count the number of characters in s first
        HashMap<Character, Integer> map = new HashMap<>();
        for(char c : s.toCharArray()){
            map.put(c, map.getOrDefault(c, 0) + 1);
            //counts the number of each letter in s
        }
        for(char ch : t.toCharArray()){
            map.put(ch, map.getOrDefault(ch, 0) - 1);
            //subtract the character counts from the map
        }
        for(int value : map.values()){
            if(value < 0 || value > 1){
                return false;
            }
        }
        return true;



    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-3.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        '''
            given string only uppercase, and integer k
            choose up to k chars of string and replace with any uppercase English cahracter

            After k replacements return length of longest substring



            EX:
                XYYX, k = 2
                We can do two replacements of the Ys with Xs or Xs with Ys
                then return length of the updated string

            EX:
                AAABABB, k = 1
                Relace the B with an A --> get 5

            
            AIM O(N) 

            Intuition:
                Similarly we need a hashmap to count unique chars
                BUT rather than just shrinking the window, we need to replace the nonunique character with whatever amkes the string the longest

                do the same max string thing -
                LIGHT WORK
        '''
        from collections import defaultdict
        l = 0
        max_freq = 0
        counts = defaultdict(int)
        max_length = 0

        for r in range(len(s)):
            #increment counts for each window
            counts[s[r]] += 1
            max_freq = max(max_freq, counts[s[r]])

            
                #if we encounter a non-unique character, we dont shrink window
                #we actually replace that character with whatever the rest of the string is

               #we dont want to modify count rather the string
               #nvm we cant  mutate strings, and either way it doesnt work
            while r - l + 1 - max_freq > k:#window size - max_freq
                #while the window size - the appearance of the most frequent char
                #left side is the number of unique chars, as long as thats > k
                counts[s[l]] -= 1
                l += 1
            max_length = max(max_length, r - l + 1)
        return max_length

               
               

